File: app/lib/models/dvd_info.dart
// DVD information models for title enumeration.
// These match the Rust worker's JSON output.

/// Complete information about a DVD disc.
class DvdInfo {
  final String volumeLabel;
  final String devicePath;
  final List<DvdTitle> titles;

  const DvdInfo({
    required this.volumeLabel,
    required this.devicePath,
    required this.titles,
  });

  factory DvdInfo.fromJson(Map<String, dynamic> json) {
    return DvdInfo(
      volumeLabel: json['volumeLabel'] as String? ?? '',
      devicePath: json['devicePath'] as String? ?? '',
      titles: (json['titles'] as List?)
              ?.map((t) => DvdTitle.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

/// Information about a single DVD title.
class DvdTitle {
  final int index;
  final double durationSeconds;
  final List<DvdChapter> chapters;
  final List<DvdAudioTrack> audioTracks;
  final int width;
  final int height;
  final double frameRate;
  final String aspectRatio;
  final int angles;
  final int vtsNumber;

  const DvdTitle({
    required this.index,
    required this.durationSeconds,
    required this.chapters,
    required this.audioTracks,
    required this.width,
    required this.height,
    required this.frameRate,
    required this.aspectRatio,
    required this.angles,
    required this.vtsNumber,
  });

  factory DvdTitle.fromJson(Map<String, dynamic> json) {
    return DvdTitle(
      index: json['index'] as int? ?? 0,
      durationSeconds: (json['durationSeconds'] as num?)?.toDouble() ?? 0.0,
      chapters: (json['chapters'] as List?)
              ?.map((c) => DvdChapter.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      audioTracks: (json['audioTracks'] as List?)
              ?.map((a) => DvdAudioTrack.fromJson(a as Map<String, dynamic>))
              .toList() ??
          [],
      width: json['width'] as int? ?? 720,
      height: json['height'] as int? ?? 480,
      frameRate: (json['frameRate'] as num?)?.toDouble() ?? 29.97,
      aspectRatio: json['aspectRatio'] as String? ?? '4:3',
      angles: json['angles'] as int? ?? 1,
      vtsNumber: json['vtsNumber'] as int? ?? 1,
    );
  }

  /// Formatted duration string (e.g., "1:23:45").
  String get durationFormatted {
    final totalSecs = durationSeconds.toInt();
    final hours = totalSecs ~/ 3600;
    final minutes = (totalSecs % 3600) ~/ 60;
    final seconds = totalSecs % 60;
    if (hours > 0) {
      return '$hours:${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';
    }
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }

  /// Resolution string (e.g., "720x480").
  String get resolution => '${width}x$height';

  /// Video system string (e.g., "NTSC" or "PAL").
  String get videoSystem => height <= 480 ? 'NTSC' : 'PAL';

  /// Audio summary string (e.g., "English AC3 5.1, French AC3 2.0").
  String get audioSummary {
    if (audioTracks.isEmpty) return 'No audio';
    return audioTracks.map((a) => a.summary).join(', ');
  }
}

/// Information about a single chapter within a title.
class DvdChapter {
  final int index;
  final double durationSeconds;

  const DvdChapter({
    required this.index,
    required this.durationSeconds,
  });

  factory DvdChapter.fromJson(Map<String, dynamic> json) {
    return DvdChapter(
      index: json['index'] as int? ?? 0,
      durationSeconds: (json['durationSeconds'] as num?)?.toDouble() ?? 0.0,
    );
  }

  /// Formatted duration string.
  String get durationFormatted {
    final totalSecs = durationSeconds.toInt();
    final minutes = totalSecs ~/ 60;
    final seconds = totalSecs % 60;
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }
}

/// Information about an audio track.
class DvdAudioTrack {
  final int index;
  final String language;
  final String format;
  final int channels;
  final int sampleRate;

  const DvdAudioTrack({
    required this.index,
    required this.language,
    required this.format,
    required this.channels,
    required this.sampleRate,
  });

  factory DvdAudioTrack.fromJson(Map<String, dynamic> json) {
    return DvdAudioTrack(
      index: json['index'] as int? ?? 0,
      language: json['language'] as String? ?? 'und',
      format: json['format'] as String? ?? 'Unknown',
      channels: json['channels'] as int? ?? 2,
      sampleRate: json['sampleRate'] as int? ?? 48000,
    );
  }

  /// Channel layout string (e.g., "5.1", "2.0").
  String get channelLayout {
    switch (channels) {
      case 1:
        return '1.0';
      case 2:
        return '2.0';
      case 6:
        return '5.1';
      case 8:
        return '7.1';
      default:
        return '$channels ch';
    }
  }

  /// Summary string (e.g., "English AC3 5.1").
  String get summary => '$language $format $channelLayout';
}

File: worker/src/models/dvd_info.rs
//! DVD information models for title enumeration.

use serde::{Deserialize, Serialize};

/// Complete information about a DVD disc.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DvdInfo {
    /// Volume label (disc name)
    pub volume_label: String,

    /// Path to the DVD mount point or device
    pub device_path: String,

    /// List of titles on the disc
    pub titles: Vec<DvdTitle>,
}

/// Information about a single DVD title.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DvdTitle {
    /// Title index (1-based)
    pub index: u32,

    /// Duration in seconds
    pub duration_seconds: f64,

    /// Chapters within this title
    pub chapters: Vec<DvdChapter>,

    /// Audio tracks
    pub audio_tracks: Vec<DvdAudioTrack>,

    /// Video width in pixels
    pub width: u32,

    /// Video height in pixels (480 for NTSC, 576 for PAL)
    pub height: u32,

    /// Frame rate (29.97 for NTSC, 25.0 for PAL)
    pub frame_rate: f64,

    /// Aspect ratio string ("4:3" or "16:9")
    pub aspect_ratio: String,

    /// Number of angles
    pub angles: u32,

    /// VTS (Video Title Set) number this title belongs to
    pub vts_number: u32,
}

/// Information about a single chapter within a title.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DvdChapter {
    /// Chapter index (1-based)
    pub index: u32,

    /// Duration in seconds (0 if unknown)
    pub duration_seconds: f64,
}

/// Information about an audio track.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DvdAudioTrack {
    /// Track index (0-based)
    pub index: u32,

    /// Language code (e.g., "en", "fr", "und" for undefined)
    pub language: String,

    /// Audio format (e.g., "AC3", "DTS", "LPCM", "MPEG-1", "MPEG-2")
    pub format: String,

    /// Number of channels (e.g., 2 for stereo, 6 for 5.1)
    pub channels: u32,

    /// Sample rate in Hz (e.g., 48000)
    pub sample_rate: u32,
}
